tests/integration/test_integration.py

```diff
@@ -46,6 +55,108 @@ async def _deploy_grafana_agent(ops_test: OpsTest):
     )
 
 
+async def _deploy_traefik(ops_test: OpsTest):
+    assert ops_test.model
+    await ops_test.model.deploy(
+        TRAEFIK_CHARM_NAME,
+        application_name=TRAEFIK_CHARM_NAME,
+        channel=TRAEFIK_CHARM_CHANNEL,
+        trust=True,
+    )
+
+
+async def configure_traefik(ops_test: OpsTest, traefik_ip: str) ->  None:
+    assert ops_test.model
+    await ops_test.model.applications[TRAEFIK_CHARM_NAME].set_config(
+        {
+            "external_hostname": f"{traefik_ip}.nip.io",
+            "routing_mode": "subdomain"
+        }
+    )
+    await ops_test.model.wait_for_idle(
+        apps=[TRAEFIK_CHARM_NAME],
+        status="active",
+        timeout=TIMEOUT,
+    )
+
+
+async def _deploy_sdcore_upf(ops_test: OpsTest):
+    assert ops_test.model
+    await ops_test.model.deploy(
+        UPF_CHARM_NAME,
+        application_name=UPF_CHARM_NAME,
+        channel=UPF_CHARM_CHANNEL,
+        trust=True,
+    )
+
+
+async def _deploy_sdcore_gnbsim(ops_test: OpsTest):
+    assert ops_test.model
+    await ops_test.model.deploy(
+        GNBSIM_CHARM_NAME,
+        application_name=GNBSIM_CHARM_NAME,
+        channel=GNBSIM_CHARM_CHANNEL,
+        trust=True,
+    )
+
+
+async def get_traefik_proxied_endpoints(ops_test: OpsTest) -> dict:
+    """Retrieve the endpoints by using Traefik's `show-proxied-endpoints` action."""
+    assert ops_test.model
+    traefik = ops_test.model.applications[TRAEFIK_CHARM_NAME]
+    traefik_unit = traefik.units[0]
+    t0 = time.time()
+    timeout = 30  # seconds
+    while time.time() - t0 < timeout:
+        proxied_endpoint_action = await traefik_unit.run_action(
+            action_name="show-proxied-endpoints"
+        )
+        action_output = await ops_test.model.get_action_output(
+            action_uuid=proxied_endpoint_action.entity_id, wait=30
+        )
+
+        if "proxied-endpoints" in action_output:
+            proxied_endpoints = json.loads(action_output["proxied-endpoints"])
+            return proxied_endpoints
+        else:
+            logger.info("Traefik did not return proxied endpoints yet")
+        time.sleep(2)
+
+    raise TimeoutError("Traefik did not return proxied endpoints")
+
+
+async def get_traefik_ip_address(ops_test: OpsTest) -> str:
+    endpoints = await get_traefik_proxied_endpoints(ops_test)
+    return _get_host_from_url(endpoints[TRAEFIK_CHARM_NAME]["url"])
+
+
+async def get_sdcore_nms_endpoint(ops_test: OpsTest) -> str:
+    endpoints = await get_traefik_proxied_endpoints(ops_test)
+    return endpoints[APP_NAME]["url"]
+
+
+def _get_host_from_url(url: str) -> str:
+    """Return the host from a URL formatted as http://<host>:<port>/ or as http://<host>/."""
+    return url.split("//")[1].split(":")[0].split("/")[0]
+
+
+def ui_is_running(nms_endpoint: str) -> bool:
+    url = f"{nms_endpoint}config/v1/network-slice"
```

Comment:
to change to `f"{nms_endpoint}/network-configuration"` once the new image is used